### PR TITLE
Add the tbl8 table settup in the factory class of ipv6

### PR DESCRIPTION
**Pull request by aarc0305**
1. Malloc memory for tbl8 table and initialize the tables.
2. Set up the rules in the tbl8 table.
3. Set up the rule in the tbl24 table and make tbl24 table point to the first
   tbl8 table.

### CODE REVIEW

**Comment by aarc0305** [comment]
@tz70s , could you review these additions for tbl8 settup? thanks!

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 1]
```diff
@@ -102,6 +118,25 @@ __global__ void SetupIPv6RuleEntry(IPv6RuleEntry *ipv6_tbl_24, unsigned long int
   }
 }
 
+__global__ void SetupIPv6TBL8RuleEntry(IPv6RuleEntry *ipv6_tbl_8, IPv6RuleEntry *next_tbl_8, unsigned long start, uint8_t next_hop, uint8_t depth) {
+  int idx = threadIdx.x;
+  if (next_tbl_8 != nullptr) {
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu]: no need to check `nullptr` here

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 2]
```diff
@@ -86,13 +95,20 @@ int IPv4LPMFactory::AddLPMRule(uint32_t ipv4_address, uint8_t depth, uint8_t nex
...
+    ipv6_tbl_24[start + idx].depth = depth;
+    ipv6_tbl_24[start + idx].tbl8_ptr = ipv6_tbl_8;
+    printf("Setup ipv6 lpm rule entry which points to the tbl8 table! index: %lu, next_hop: %d\n", (start + idx), next_hop); 
   } else {
-    // Add the new rule
-    // TODO: Currently, we only handle with the case of depth <= 24. The case of depth > 24 using the tbl8 table has not been implemented.
+    // Add the new rule but TBL24 does not have to point to the tbl8 table.
     ipv6_tbl_24[start + idx].next_hop = next_hop;
     ipv6_tbl_24[start + idx].valid_flag = true;
     ipv6_tbl_24[start + idx].external_flag = false;
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu]: Eliminating code.
```c++
 // Add the new rule and TBL24 has to point to the tbl8 table
ipv6_tbl_24[start + idx].next_hop = next_hop;
ipv6_tbl_24[start + idx].valid_flag = true;
ipv6_tbl_24[start + idx].depth = depth;
ipv6_tbl_24[start + idx].tbl8_ptr = ipv6_tbl_8;
if (ipv6_tbl_8 != nullptr) ipv6_tbl_24[start + idx].external_flag = true;
```

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 3]
```diff
@@ -147,18 +185,49 @@ int IPv6LPMFactory::AddLPMRule(uint8_t *ipv6_address, uint8_t depth, uint8_t nex
     start = right_shift << (24 - depth);
       
   } else {
-    // TODO: Handling when depth is larger than 24
+    // Handling when depth is larger than 24
     start = ipv6_address[0] << 16 | ipv6_address[1] << 8 | ipv6_address[2];
+   
+    // Calculate the number of tbl8 table 
+    int tbl_8_number = depth % 8 ? ((depth - 24) / 8) + 1 : (depth - 24) / 8;
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu:174]: Looks not correct, consider depth = 25?

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu (hunk 2 above)
Comment (aarc0305) [reply gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu]: ok, I would modify it later.

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu (hunk 3 above)
Comment (aarc0305) [reply gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu:174]: When depth = 25, there would be 1 tbl8 table?

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 4]
```diff
@@ -110,10 +145,13 @@ int IPv6LPMFactory::AddLPMRule(uint8_t *ipv6_address, uint8_t depth, uint8_t nex
...
   unsigned long int distance = 1;
-  for (unsigned long int i = 0; i < (24 - depth); i++) {
-    distance *= 2;
+  if (depth <= 24) {
+    for (unsigned long int i = 0; i < (24 - depth); i++) {
+      distance *= 2;
+    }
   }
 
+  IPv6RuleEntry **ipv6_tbl_8_array = nullptr;
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu:136]: Move down after the distance counting, where you allocating pointers.

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 5]
```diff
@@ -147,18 +185,49 @@ int IPv6LPMFactory::AddLPMRule(uint8_t *ipv6_address, uint8_t depth, uint8_t nex
...
+    int tbl_8_number = depth % 8 ? ((depth - 24) / 8) + 1 : (depth - 24) / 8;
+    
+    // Calculate the distance of the last tbl8 table. 
+    int tbl_8_last_table_distance = 1;
+    for (int i = 0; i < (depth % 8); i++) {
+      tbl_8_last_table_distance *= 2;
+    }
+   
+    // Malloc memory in the cuda device,  initialize them and set up the rules. 
+    ipv6_tbl_8_array = new IPv6RuleEntry *[tbl_8_number];
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu:183]: Move here.

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 6]
```diff
@@ -147,18 +185,49 @@ int IPv6LPMFactory::AddLPMRule(uint8_t *ipv6_address, uint8_t depth, uint8_t nex
...
 
   unsigned long num_of_threads = 1024;
   if (distance <= num_of_threads) {
-    SetupIPv6RuleEntry<<<1, distance>>>(IPv6TBL24, start, next_hop, depth); 
+    if (ipv6_tbl_8_array == nullptr) {
+      // TBL24 does not have to point to the tbl8 table.
+      SetupIPv6RuleEntry<<<1, distance>>>(IPv6TBL24, start, next_hop, depth, nullptr);
+    } else {
+      // TBL24 has to point to the tbl8 table.
+      SetupIPv6RuleEntry<<<1, distance>>>(IPv6TBL24, start, next_hop, depth, ipv6_tbl_8_array[0]);
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu:207]: Again, you don't have to pass them with nullptr check here.

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu (hunk 3 above)
Comment (tz70s) [reply gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu:174]: But `(25-24) / 8 == 0`?

Comment (tz70s) [reply gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu:174]: The correct version is 
```c++
int num_of_tbl8 = (depth - 24) / 8 + 1;
```

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 7]
```diff
@@ -40,6 +40,15 @@ __global__ void InitIPv6LPMTable(IPv6RuleEntry *ipv6_tbl_24) {
...
 
+__global__ void InitIPv6TBL8Table(IPv6RuleEntry *ipv6_tbl_8) {
+  int idx = threadIdx.x;
+  ipv6_tbl_8[idx].next_hop = 254;
+  ipv6_tbl_8[idx].valid_flag = false;
+  ipv6_tbl_8[idx].depth = 0;
+  ipv6_tbl_8[idx].external_flag = false;
+  ipv6_tbl_8[idx].tbl8_ptr = nullptr;
+}
+
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu:51]: Also, to improve the readability, it's time to use template to generate these repetition **intialization** and **setups** **kernel functions**. The type will be inferred when calling kernel functions.

```c++
template <typename IPvxRuleEntry>
__global__ void InitIPvxRuleEntry(IPvxRuleEntry *ipvx_tbl_ptr) {
  int idx = blockIdx.x * blockDim.x + threadIdx.x;
  ipvx_tbl_ptr[idx].next_hop = 254;
  ipvx_tbl_ptr[idx].valid_flag = false;
  ipvx_tbl_ptr[idx].depth = 0;
  ipvx_tbl_ptr[idx].external_flag = false;
  ipvx_tbl_ptr[idx].tbl8_ptr = nullptr;
}
```

Comment (tz70s) [reply gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu:51]: To Improve consistency, you need to add the tbl8_ptr field on IPv4RuleEntry, but need not to implement it in detailed. ( Or using template specialized, but not recommend here. )

**Review by tz70s — CHANGES_REQUESTED** [review]
Basically, the functionalities looks correct, but the code is getting messy and redundant.

**Comment by aarc0305** [comment]
@tz70s , could you please review these changes? Thanks!

**Review by tz70s — APPROVED** [review]

**Comment by tz70s** [comment]
Merge first, I'll fix up remaining issues.

**Pull request merged**